icicle/src/hash/blake2s/blake2s.cu

```diff
@@ -0,0 +1,278 @@
+        memcpy(ctx->buff + ctx->pos, in, start);
+        ctx->t0 += BLAKE2S_BLOCK_LENGTH;
+
+        if (ctx->t0 == 0) ctx->t1++;
+
+        cuda_blake2s_compress(ctx, ctx->buff, 0);
+        ctx->pos = 0;
+        memset(ctx->buff, 0, BLAKE2S_BLOCK_LENGTH);
+      } else {
+        memcpy(ctx->buff + ctx->pos, in, inlen);
+        ctx->pos += inlen;
+        return;
+      }
+    }
+
+    block_index = inlen - BLAKE2S_BLOCK_LENGTH;
+    for (in_index = start; in_index < block_index; in_index += BLAKE2S_BLOCK_LENGTH) {
+      ctx->t0 += BLAKE2S_BLOCK_LENGTH;
+      if (ctx->t0 == 0) ctx->t1++;
+
+      cuda_blake2s_compress(ctx, in, in_index);
+    }
+
+    memcpy(ctx->buff, in + in_index, inlen - in_index);
+    ctx->pos += inlen - in_index;
+  }
+
+  __device__ void cuda_blake2s_final(cuda_blake2s_ctx_t* ctx, BYTE* out)
+  {
+    ctx->f0 = 0xFFFFFFFFUL;
+    ctx->t0 += ctx->pos;
+    if (ctx->pos > 0 && ctx->t0 == 0) ctx->t1++;
+
+    cuda_blake2s_compress(ctx, ctx->buff, 0);
+    memset(ctx->buff, 0, BLAKE2S_BLOCK_LENGTH);
+    memset(ctx->state, 0, BLAKE2S_STATE_LENGTH);
+
+    int i4 = 0;
+    for (int i = 0; i < BLAKE2S_CHAIN_SIZE && ((i4 = i * 4) < ctx->digestlen); i++) {
+      BYTE* BYTEs = (BYTE*)(&ctx->chain[i]);
+      if (i4 < ctx->digestlen - 4)
+        memcpy(out + i4, BYTEs, 4);
+      else
+        memcpy(out + i4, BYTEs, ctx->digestlen - i4);
+    }
+  }
+
+  __global__ void
+  kernel_blake2s_hash(const BYTE* indata, WORD inlen, BYTE* outdata, WORD n_batch, WORD BLAKE2S_BLOCK_SIZE)
+  {
+    WORD thread = blockIdx.x * blockDim.x + threadIdx.x;
+    if (thread >= n_batch) { return; }
+    BYTE key[32] = ""; // Null key
+    WORD keylen = 0;
+    CUDA_BLAKE2S_CTX blake_ctx;
+    const BYTE* in = indata + thread * inlen;
+    BYTE* out = outdata + thread * BLAKE2S_BLOCK_SIZE;
+
+    cuda_blake2s_init(&blake_ctx, key, keylen, (BLAKE2S_BLOCK_SIZE << 3));
+    cuda_blake2s_update(&blake_ctx, in, inlen);
+    cuda_blake2s_final(&blake_ctx, out);
+  }
+
+  extern "C" {
+  void
+  mcm_cuda_blake2s_hash_batch(BYTE* key, WORD keylen, BYTE* in, WORD inlen, BYTE* out, WORD output_len, WORD n_batch)
+  {
+    BYTE* cuda_indata;
+    BYTE* cuda_outdata;
+    const WORD BLAKE2S_BLOCK_SIZE = output_len;
+    cudaMalloc(&cuda_indata, inlen * n_batch);
+    cudaMalloc(&cuda_outdata, BLAKE2S_BLOCK_SIZE * n_batch);
+    assert(keylen <= 32);
+
+    // CUDA_BLAKE2S_CTX ctx;
+    // cpu_blake2s_init(&ctx, key, keylen, n_outbit);
+    // cudaMemcpyToSymbol(c_CTX, &ctx, sizeof(CUDA_BLAKE2S_CTX), 0, cudaMemcpyHostToDevice);
+
+    cudaMemcpy(cuda_indata, in, inlen * n_batch, cudaMemcpyHostToDevice);
+
+    WORD thread = 256;
+    WORD block = (n_batch + thread - 1) / thread;
+    kernel_blake2s_hash<<<block, thread>>>(cuda_indata, inlen, cuda_outdata, n_batch, BLAKE2S_BLOCK_SIZE);
+    cudaMemcpy(out, cuda_outdata, BLAKE2S_BLOCK_SIZE * n_batch, cudaMemcpyDeviceToHost);
+    cudaDeviceSynchronize();
+    cudaError_t error = cudaGetLastError();
```

Comment:
Please use our error-management functions (you can find an example in any of our primitives)